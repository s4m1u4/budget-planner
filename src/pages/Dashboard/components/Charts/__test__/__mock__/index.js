export { CATEGORIES, RECORDS } from "./mock";
