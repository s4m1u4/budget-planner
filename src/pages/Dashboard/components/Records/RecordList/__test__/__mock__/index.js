export { LAST_RECORDS } from "./mock";
