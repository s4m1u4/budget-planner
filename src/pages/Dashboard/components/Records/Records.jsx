import React, { useState } from "react";
import { IconButton } from "@mui/material";
import { RecordsForm } from "./RecordsForm";
import { RecordsList } from "./RecordList";
import AddCircleOutlineSharpIcon from "@mui/icons-material/AddCircleOutlineSharp";

import {
  Section,
  SectionHeader,
  SectionFooter,
  SectionTitle,
} from "../../Dashboard.styles";

export const Records = ({ lastRecords, setNewHistory, categories }) => {
  const [isOpenModal, setIsOpenModal] = useState(false);
  const handleOpen = () => setIsOpenModal(true);
  const handleClose = () => setIsOpenModal(false);

  return (
    <Section area="records">
      <SectionHeader>
        <SectionTitle>Last records</SectionTitle>
        <IconButton
          size="small"
          onClick={handleOpen}
          sx={{ padding: "0" }}
          disabled={!categories.length}
        >
          <AddCircleOutlineSharpIcon />
        </IconButton>
      </SectionHeader>
      <SectionFooter>
        {lastRecords.length ? (
          <RecordsList lastRecords={lastRecords} />
        ) : (
          <p>Records list is emptyðŸ˜”</p>
        )}
      </SectionFooter>
      <RecordsForm
        open={isOpenModal}
        categories={categories}
        handleClose={handleClose}
        setNewHistory={setNewHistory}
      />
    </Section>
  );
};
