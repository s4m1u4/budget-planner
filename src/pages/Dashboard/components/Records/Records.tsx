import React, { FC, useState } from "react";
import { IconButton } from "@mui/material";
import { RecordsForm } from "./RecordsForm";
import { RecordsList } from "./RecordList";
import AddCircleOutlineSharpIcon from "@mui/icons-material/AddCircleOutlineSharp";
import { ICategory, IRecord } from "../../../../types";
import { IHistoryData } from "../../types";

import {
  Section,
  SectionHeader,
  SectionFooter,
  SectionTitle,
} from "../../Dashboard.styles";

interface RecordsProps {
  lastRecords: IRecord[];
  categories: ICategory[];
  setNewHistory: (historyData: IHistoryData) => void;
}

export const Records: FC<RecordsProps> = ({
  lastRecords,
  setNewHistory,
  categories,
}) => {
  const [isOpenModal, setIsOpenModal] = useState<boolean>(false);
  const handleOpen = () => setIsOpenModal(true);
  const handleClose = () => setIsOpenModal(false);

  return (
    <Section area="records">
      <SectionHeader>
        <SectionTitle>Last records</SectionTitle>
        <IconButton
          size="small"
          data-modal="modal"
          onClick={handleOpen}
          sx={{ padding: "0" }}
          disabled={!categories.length}
        >
          <AddCircleOutlineSharpIcon />
        </IconButton>
      </SectionHeader>
      <SectionFooter>
        {lastRecords.length ? (
          <RecordsList lastRecords={lastRecords} />
        ) : (
          <p>Records list is emptyðŸ˜”</p>
        )}
      </SectionFooter>
      <RecordsForm
        open={isOpenModal}
        onSubmit={(values) => values}
        categories={categories}
        handleClose={handleClose}
        setNewHistory={setNewHistory}
      />
    </Section>
  );
};
