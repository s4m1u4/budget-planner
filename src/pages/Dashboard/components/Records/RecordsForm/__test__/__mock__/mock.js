export const CATEGORIES = [
  {
    id: "084a3558-6b14-4e38-9d24-47e912021384",
    title: "Work",
    description: "Work",
    isDefault: false,
  },
  {
    id: "0ce314b9-a547-4f57-9d64-e466777ba8e9",
    title: "Travel",
    description: "awdawda",
    isDefault: false,
  },
  {
    id: "ea298d16-8bdb-4c5b-8198-6b2e5ce2f9a2",
    title: "Education",
    description: "Education it's so important for you!",
    isDefault: false,
  },
];
