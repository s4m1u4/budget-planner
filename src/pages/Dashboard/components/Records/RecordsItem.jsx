import React from "react";
import currency from "currency.js";
import { Payment, PaymentTitle, PaymentTotal } from "../../Dashboard.styles";

export const RecordsItem = ({ record }) => {
  return (
    <Payment key={record.id}>
      <PaymentTitle>
        {record.type} from {record.category.title}
      </PaymentTitle>
      <PaymentTotal color={record.type}>
        {record.type === "income"
          ? currency(record.amount, { pattern: `+!#` }).format()
          : currency(record.amount, { pattern: `-!#` }).format()}
      </PaymentTotal>
    </Payment>
  );
};
