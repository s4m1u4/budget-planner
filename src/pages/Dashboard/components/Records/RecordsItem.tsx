import React, { FC } from "react";
import currency from "currency.js";
import { IRecord } from "../../../../types";

import { Payment, PaymentTitle, PaymentTotal } from "../../Dashboard.styles";

interface RecordsItemProps {
  record: IRecord;
}

export const RecordsItem: FC<RecordsItemProps> = ({ record }) => {
  return (
    <Payment key={record.id}>
      <PaymentTitle>
        {record.type} from {record.category?.title}
      </PaymentTitle>
      <PaymentTotal color={record.type}>
        {record.type === "income"
          ? currency(record.amount, { pattern: `+!#` }).format()
          : currency(record.amount, { pattern: `-!#` }).format()}
      </PaymentTotal>
    </Payment>
  );
};
