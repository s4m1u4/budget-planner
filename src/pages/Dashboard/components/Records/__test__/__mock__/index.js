export { CATEGORIES, LAST_RECORDS } from "./mock";
