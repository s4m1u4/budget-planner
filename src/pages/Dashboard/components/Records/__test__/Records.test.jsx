import { Records } from "../Records";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { CATEGORIES, LAST_RECORDS } from "./__mock__";

const renderComponent = (lastRecords) => {
  return render(
    <Records
      categories={CATEGORIES}
      lastRecords={lastRecords}
      setNewHistory={setNewHistory}
    />
  );
};

const setNewHistory = jest.fn();

describe("Records component", () => {
  describe("with empty records", () => {
    it("snapshot matching", () => {
      const { baseElement } = renderComponent([]);
      expect(baseElement).toMatchSnapshot();
    });

    describe("should render", () => {
      beforeEach(() => {
        renderComponent([]);
      });

      it("the title", () => {
        expect(
          screen.getByRole("heading", { name: /last records/i })
        ).toBeInTheDocument();
      });

      it("the text 'Records list is empty'", () => {
        expect(screen.getByText(/records list is empty/i)).toBeInTheDocument();
      });
    });
  });

  describe("with list of records", () => {
    it("snapshot matching", () => {
      const { baseElement } = renderComponent(LAST_RECORDS);
      expect(baseElement).toMatchSnapshot();
    });

    describe('modal "Create a new record"', () => {
      it("should open", () => {
        const { container } = renderComponent(LAST_RECORDS);
        const buttonOpenModal = container.querySelector('[data-modal="modal"]');

        userEvent.click(buttonOpenModal);

        expect(
          screen.getByRole("heading", { name: /create a new record/i })
        ).toBeInTheDocument();
      });

      it("should close", () => {
        const { container } = renderComponent(LAST_RECORDS);
        const buttonOpenModal = container.querySelector('[data-modal="modal"]');

        userEvent.click(buttonOpenModal);

        expect(
          screen.getByRole("heading", { name: /create a new record/i })
        ).toBeInTheDocument();

        userEvent.click(screen.getByRole("button", { name: /close/i }));

        expect(
          screen.queryByRole("heading", { name: /create a new record/i })
        ).not.toBeInTheDocument();
      });
    });
  });
});
