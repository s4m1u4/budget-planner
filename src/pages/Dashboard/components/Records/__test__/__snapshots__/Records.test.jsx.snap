// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Records component with empty records snapshot matching 1`] = `
<body>
  <div>
    <div
      class="css-sai3sf"
    >
      <div
        class="css-13o2qpx"
      >
        <h3
          class="css-1xqtn9i"
        >
          Last records
        </h3>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1mgsdk9-MuiButtonBase-root-MuiIconButton-root"
          data-modal="modal"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AddCircleOutlineSharpIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="css-1gk4ta0"
      >
        <p>
          Records list is emptyðŸ˜”
        </p>
      </div>
    </div>
  </div>
</body>
`;

exports[`Records component with list of records snapshot matching 1`] = `
<body>
  <div>
    <div
      class="css-sai3sf"
    >
      <div
        class="css-13o2qpx"
      >
        <h3
          class="css-1xqtn9i"
        >
          Last records
        </h3>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1mgsdk9-MuiButtonBase-root-MuiIconButton-root"
          data-modal="modal"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AddCircleOutlineSharpIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="css-1gk4ta0"
      >
        <div
          class="css-7wb1st"
        >
          <div
            class="css-gg4vpm"
          >
            <p
              class="css-0"
            >
              income
               from 
            </p>
            <span
              class="css-81iyw9"
              color="income"
            >
              +$2,354.00
            </span>
          </div>
          <div
            class="css-gg4vpm"
          >
            <p
              class="css-0"
            >
              income
               from 
            </p>
            <span
              class="css-81iyw9"
              color="income"
            >
              +$836.00
            </span>
          </div>
          <div
            class="css-gg4vpm"
          >
            <p
              class="css-0"
            >
              income
               from 
            </p>
            <span
              class="css-81iyw9"
              color="income"
            >
              +$2,454.00
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
